rust/transaction-filter/src/filter_operator.rs

```diff
@@ -0,0 +1,336 @@
+// use crate::traits::Filterable;
+use crate::{
+    filters::{
+        EventFilter, TransactionRootFilter, UserTransactionRequestFilter, WriteSetChangeFilter,
+    },
+    traits::Filterable,
+};
+use anyhow::Error;
+use aptos_protos::transaction::v1::{transaction::TxnData, Transaction};
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+
+/// These are filters we would expect to be exposed via API
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+#[serde(tag = "type")]
+pub enum APIFilter {
+    TransactionRootFilter(TransactionRootFilter),
+    UserTransactionRequestFilter(UserTransactionRequestFilter),
+    EventFilter(EventFilter),
+    WriteSetChangeFilter(WriteSetChangeFilter),
+}
+
+impl Filterable<Transaction> for APIFilter {
+    fn is_valid(&self) -> Result<(), Error> {
+        match self {
+            APIFilter::TransactionRootFilter(filter) => filter.is_valid(),
+            APIFilter::UserTransactionRequestFilter(filter) => filter.is_valid(),
+            APIFilter::EventFilter(filter) => filter.is_valid(),
+            APIFilter::WriteSetChangeFilter(filter) => filter.is_valid(),
+        }
+    }
+
+    fn is_allowed(&self, txn: &Transaction) -> bool {
+        match self {
+            APIFilter::TransactionRootFilter(filter) => filter.is_allowed(txn),
+            APIFilter::UserTransactionRequestFilter(ut_filter) => txn
+                .txn_data
+                .as_ref()
+                .map(|txn_data| {
+                    if let TxnData::User(u) = txn_data {
+                        u.request
+                            .as_ref()
+                            .map(|req| ut_filter.is_allowed(req))
+                            .unwrap_or(false)
+                    } else {
+                        false
+                    }
+                })
+                .unwrap_or(false),
```

Comment:
Nit, if possible it'd be cleaner if this logic was just inside UserTransactionRequestFilter.is_allowed.